src/misc/umap.c

```diff
@@ -0,0 +1,846 @@
+ *       different depending on the point of view/angle, and there are a lot of
+ *       viewpoints when the dimensionality ramps up.
+ *
+ * </para><para>
+ * References:
+ *
+ * </para><para>
+ * Leland McInnes, John Healy, and James Melville, arXiv:1802.03426v3
+ *
+ * \param graph Pointer to the similarity graph to find a layout for (i.e. to embed). Weights in
+ *   this graph object are currently ignored.
+ * \param distances Pointer to a vector of "distances" between vertices. Similarity graphs for
+ *   UMAP are often originally meant in terms of similarity weights (e.g. correlation between
+ *   high-dimensional vectors) and converted into distances by crude dist := 1 - corr. That is
+ *   fine here too. If this argument is an empty pointer (NULL), connected vertices are assumed
+ *   to be very similar.
+ * \param layout Pointer to the n x 2 matrix where the layout coordinates will be stored. Only 2D
+ *   embedding are currently supported, as they are by far more common than any higher dimensions.
+ * \param min_dist A fudge parameter that decides how close two unconnected vertices can be in the
+ *   embedding before feeling a repulsive force. Typically, 0.01 is a good number. If this is
+ *   negative or zero, 0.01 is assumed.
+ * \param epochs Number of iterations of the main stochastic gradient descent loop on the
+ *   cross-entropy. If negative or zero, 500 epochs are used if the graph is the graph is small
+ *   (less than 50k edges), 50 epochs are used for larger graphs.
+ * \param sampling_prob The fraction of vertices moved at each iteration of the stochastic gradient descent (epoch). At fixed number of epochs, a higher fraction makes the algorithm slower. Vice versa, a too low number will converge very slowly, possibly too slowly. If negative of zero, a value of 0.3 is assumed.
+ *
+ * \return Error code.
+ *
+ */
+igraph_error_t igraph_layout_umap(const igraph_t *graph, const igraph_vector_t *distances,
+    igraph_matrix_t *layout, const igraph_real_t min_dist, const igraph_integer_t epochs, const igraph_real_t sampling_prob);
+
+
+/* rho is just the size of the distance from each vertex and its closest neighbor */
+/* sigma is the the decay from each vertex, depends on its rho and the rest of its neighbor
+ * distances */
+
+/* Find sigma for this vertex by binary search */
+static igraph_error_t igraph_i_umap_find_sigma(const igraph_t *graph,
+        const igraph_vector_t *distances, igraph_integer_t i, const igraph_vector_int_t *eids,
+        igraph_real_t rho, igraph_real_t *sigma_p, igraph_real_t target) {
+
+    igraph_real_t sigma = 1;
+    igraph_real_t sum;
+    igraph_real_t tol = 0.01;
+    igraph_integer_t maxiter = 100;
+    igraph_integer_t no_of_neis = igraph_vector_int_size(eids);
+    igraph_integer_t eid;
+    igraph_real_t step = sigma;
+    int seen_max = 0;
+
+    /* Binary search */
+    for(int iter = 0; iter < maxiter; iter++) {
```

Comment:
All variables used to index vectors should be `igraph_integer_t`. Is there any reason this is an `int`?  Whenever I see an `int` or a `long`, I wonder if this is a bug, i.e. a failure to change a type during the integer transition process. When there is a reason to keep `int`, add a comment and explain why.

Formatting:Use `for (`, not `for(`, please.